File: src.net/Marketo-Rest/Retry.cs
using Marketo.Require;
using Newtonsoft.Json.Linq;
using System;
using System.Net;
using System.Threading.Tasks;

namespace Marketo
{
    internal class Retry
    {
        const int DefaultRetryCount = 5;
        const int DefaultInitialDelay = 1000; // 1 second
        const int DefaultMaxDelay = 20000;
        readonly Action<string> _log = util.logger;
        readonly dynamic _options;
        readonly Backoff _newBackoff;

        public Retry(dynamic options)
        {
            _options = options ?? new { };
            _newBackoff = NewBackoff();
        }

        int NumRetries { get { return dyn.getProp(_options, "maxRetries", DefaultRetryCount); } }
        int InitialDelay { get { return dyn.getProp(_options, "initialDelay", DefaultInitialDelay); } }
        int MaxDelay { get { return dyn.getProp(_options, "maxDelay", DefaultMaxDelay); } }

        Exception _lastError;
        internal Task<JObject> start(Func<bool, Task<JObject>> fetchFn)
        {
            var promise = new TaskCompletionSource<JObject>();
            var forceOAuth = false;
            executeFn();
            _newBackoff.OnReady = (number, delay) => { _log($"Requesting for url {{ number: {number}, delay: {delay} }}"); executeFn(); };
            _newBackoff.OnFail = (err) => throw _lastError;
            async void executeFn()
            {
                try
                {
                    var result = await fetchFn(forceOAuth);
                    promise.SetResult(result);
                }
                catch (Exception err)
                {
                    _lastError = err;
                    if (RetryableError(err))
                    {
                        if (errors.isRateLimited(err)) await Task.Delay(MaxDelay);
                        else if (errors.isExpiredToken(err)) forceOAuth = true;
                        await _newBackoff.DoBackoff();
                        return;
                    }
                    var e = (err as RestlerOperationException);
                    if (e != null && e.Content != null && e.Content is JObject && ((JObject)e.Content)["errors"] != null)
                        err = e.StatusCode == HttpStatusCode.Unauthorized || e.StatusCode == HttpStatusCode.Forbidden ?
                            new MarketoSecurityException(e.StatusCode, ((JObject)e.Content)["errors"]) :
                            new MarketoException(e.StatusCode, ((JObject)e.Content)["errors"]);
                    promise.SetException(err);
                    throw err;
                }
            }
            return promise.Task;
        }

        bool RetryableError(Exception err)
        {
            return errors.isNetworkError(err) ||
              errors.isExpiredToken(err) ||
              errors.isServerError(err) ||
              errors.isRateLimited(err) ||
              false;
        }

        Backoff NewBackoff()
        {
            var newBackoff = new Backoff(new ExponentialBackoffStrategy(
                randomisationFactor: 0,
                initialDelay: InitialDelay,
                maxDelay: MaxDelay
            ));
            newBackoff.FailAfter(NumRetries);
            return newBackoff;
        }
    }
}

File: lib/retry.js
var _ = require('lodash'),
  backoff = require('backoff'),
  Promise = require('bluebird'),
  errors = require('./errors'),
  util = require('./util'),
  log = util.logger();

function Retry(options) {
  this._options = options || {};
}

Retry.prototype = {
  defaultRetryCount: 5,
  defaultInitialDelay: 1000, // 1 second
  defaultMaxDelay: 20000,

  numRetries: function () {
    return this._options.maxRetries || this.defaultRetryCount;
  },

  initialDelay: function () {
    return this._options.initialDelay || this.defaultInitialDelay;
  },

  maxDelay: function () {
    return this._options.maxDelay || this.defaultMaxDelay;
  },

  // Starts the retry loop, you can pass in an optional context for the
  // passed in function to be in.
  start: function (fetchFn, opt_context) {
    var self = this,
      def = Promise.defer(),
      expBackoff = this._newBackoff(),
      forceOAuth = false,
      lastError;

    if (_.isObject(opt_context)) {
      fetchFn = _.bind(fetchFn, opt_context);
    }

    executeFn();
    expBackoff.on('ready', function (number, delay) {
      log.debug('Requesting for url', { number: number, delay: delay });
      executeFn();
    });

    expBackoff.on('fail', function () {
      def.reject(lastError);
    });

    function executeFn() {
      fetchFn(forceOAuth)
        .then(function (result) {
          def.resolve(result);
        })
        .catch(function (err) {
          lastError = err;
          if (self.retryableError(err)) {
            if (errors.isRateLimited(err)) {
              // HACK: there's no way to custom set the next delay in node-backoff
              // so we'll have to do a manual timeout here
              setTimeout(function () {
                expBackoff.backoff();
              }, self.maxDelay());
            } else {
              if (errors.isExpiredToken(err)) {
                forceOAuth = true;
              }
              expBackoff.backoff();
            }
          } else {
            def.reject(err);
          }
        });
    }

    return def.promise;
  },

  retryableError: function (err) {
    return (
      errors.isAwaitable(err) ||
      errors.isNetworkError(err) ||
      errors.isExpiredToken(err) ||
      errors.isServerError(err) ||
      errors.isRateLimited(err) ||
      false
    );
  },

  _newBackoff: function () {
    var expBackoff = backoff.exponential({
      randomisationFactor: 0,
      initialDelay: this.initialDelay(),
      maxDelay: this.maxDelay(),
    });
    expBackoff.failAfter(this.numRetries());

    return expBackoff;
  },
};

module.exports = Retry;
